sys/kern/kcsan.c

```diff
@@ -0,0 +1,262 @@
+  *size = (watchpoint & WATCHPOINT_SIZE_MASK) >> WATCHPOINT_SIZE_SHIFT;
+  *is_read = (watchpoint & WATCHPOINT_READ_BIT) == WATCHPOINT_READ_BIT;
+  return true;
+}
+
+static inline bool address_match(uintptr_t addr1, size_t size1, uintptr_t addr2,
+                                 size_t size2) {
+  return (addr1 <= addr2 + size2 - 1) && (addr2 <= addr1 + size1 - 1);
+}
+
+static inline int watchpoint_slot(uintptr_t addr) {
+  return (addr / PAGESIZE) % WATCHPOINT_NUM;
+}
+
+static inline bool should_watch(void) {
+  return atomic_fetch_sub(&skip_counter, 1) == 1;
+}
+
+static void delay(int count) {
+  for (int i = 0; i < count; i++) {
+    __asm__ volatile("nop");
+  }
+}
+
+static inline atomic_int *find_watchpoint(uintptr_t addr, size_t size,
+                                          bool is_read) {
+  uintptr_t other_addr;
+  size_t other_size;
+  bool other_is_read;
+
+  int slot = watchpoint_slot(addr);
+  atomic_int *watchpoint_p = &watchpoints[slot];
+
+  int encoded = atomic_load(watchpoint_p);
+  if (!decode_watchpoint(encoded, &other_addr, &other_size, &other_is_read))
+    return NULL;
+
+  /*
+   * Data race occurs when there are two memory accesses and at least one of
+   * them is write.
+   */
+  if (other_is_read && is_read)
+    return NULL;
+
+  if (address_match(addr, size, other_addr, other_size))
+    return watchpoint_p;
+
+  return NULL;
+}
+
+static inline atomic_int *insert_watchpoint(uintptr_t addr, size_t size,
+                                            bool is_read) {
+  int slot = watchpoint_slot(addr);
+  atomic_int *watchpoint_p = &watchpoints[slot];
+
+  int encoded = encode_watchpoint(addr, size, is_read);
+  int expected = WATCHPOINT_INVALID;
+  if (!atomic_compare_exchange_strong(watchpoint_p, &expected, encoded))
+    return NULL;
+  return watchpoint_p;
+}
+
+static inline void setup_watchpoint(uintptr_t addr, size_t size, bool is_read) {
+  union {
+    uint8_t b1;
+    uint16_t b2;
+    uint32_t b4;
+    uint64_t b8;
+  } prev_val = {}, new_val = {};
+
+  /* Reset the counter. */
+  atomic_store(&skip_counter, SKIP_COUNT);
+
+  atomic_int *watchpoint_p = insert_watchpoint(addr, size, is_read);
+  if (watchpoint_p == NULL) {
+    atomic_fetch_add(&kcsan_slot_taken_count, 1);
+    return;
+  }
+
+  atomic_fetch_add(&kcsan_set_watchpoint_count, 1);
+
+  switch (size) {
+    case 1:
+      prev_val.b1 = *((uint8_t *)addr);
+      break;
+    case 2:
+      prev_val.b2 = *((uint16_t *)addr);
+      break;
+    case 4:
+      prev_val.b4 = *((uint32_t *)addr);
+      break;
+    case 8:
+      prev_val.b8 = *((uint64_t *)addr);
+      break;
+    default:
+      break;
+  }
+
+  delay(WATCHPOINT_DELAY);
+
+  switch (size) {
+    case 1:
+      new_val.b1 = *((uint8_t *)addr);
+      break;
+    case 2:
+      new_val.b2 = *((uint16_t *)addr);
+      break;
+    case 4:
+      new_val.b4 = *((uint32_t *)addr);
+      break;
+    case 8:
+      new_val.b8 = *((uint64_t *)addr);
+      break;
+    default:
```

Comment:
You can consider moving the switch to a separate function so that it's not duplicated here. Not sure if worth the effort, though.

Reply:
It turned out the effort wasn't too big, so I've changed it.